src/deluge/gui/views/performance_session_view.cpp

```diff
@@ -0,0 +1,2045 @@
+//if param is found, it loads the layout info for that param into the view (paramKind, paramID, xDisplay, yDisplay, rowColour, rowTailColour)
+void PerformanceSessionView::readDefaultFXParamFromFile(int32_t xDisplay) {
+	char const* paramName;
+	char const* tagName = storageManager.readTagOrAttributeValue();
+
+	for (int32_t i = 0; i < kNumParamsForPerformance; i++) {
+		if (songParamsForPerformance[i].paramKind == Param::Kind::UNPATCHED_GLOBAL) {
+			paramName = GlobalEffectable::paramToString(Param::Unpatched::START + songParamsForPerformance[i].paramID);
+		}
+		else if (songParamsForPerformance[i].paramKind == Param::Kind::UNPATCHED_SOUND) {
+			paramName =
+			    ModControllableAudio::paramToString(Param::Unpatched::START + songParamsForPerformance[i].paramID);
+		}
+		if (!strcmp(tagName, paramName)) {
+			memcpy(&layoutForPerformance[xDisplay], &songParamsForPerformance[i], sizeParamsForPerformance);
+
+			memcpy(&backupXMLDefaultLayoutForPerformance[xDisplay], &layoutForPerformance[xDisplay],
+			       sizeParamsForPerformance);
+			break;
+		}
+	}
+}
+
+void PerformanceSessionView::readDefaultFXRowNumberValuesFromFile(int32_t xDisplay) {
+	char const* tagName;
+	char rowNumber[5];
+	//loop through all row <#> number tags
+	while (*(tagName = storageManager.readNextTagOrAttributeName())) {
+		//find the row number that the tag corresponds to
+		//reads from row 8 down to row 1
+		for (int32_t yDisplay = kDisplayHeight - 1; yDisplay >= 0; yDisplay--) {
+			intToString(yDisplay + 1, rowNumber);
+			if (!strcmp(tagName, rowNumber)) {
+				defaultFXValues[xDisplay][yDisplay] = storageManager.readTagOrAttributeValueInt() - kKnobPosOffset;
+
+				//check if a value greater than 64 was entered as a default value in xml file
+				if (defaultFXValues[xDisplay][yDisplay] > kKnobPosOffset) {
+					defaultFXValues[xDisplay][yDisplay] = kKnobPosOffset;
+				}
+
+				backupXMLDefaultFXValues[xDisplay][yDisplay] = defaultFXValues[xDisplay][yDisplay];
+
+				break;
+			}
+		}
+		storageManager.exitTag();
+	}
+}
+
+void PerformanceSessionView::readDefaultFXHoldStatusFromFile(int32_t xDisplay) {
+	char const* tagName;
+	//loop through the hold tags
+	while (*(tagName = storageManager.readNextTagOrAttributeName())) {
+		//<status>
+		if (!strcmp(tagName, STRING_FOR_PERFORM_DEFAULTS_HOLD_STATUS_TAG)) {
+			char const* holdStatus = storageManager.readTagOrAttributeValue();
+			if (!strcmp(holdStatus, l10n::get(l10n::String::STRING_FOR_ON))) {
+				if (!isParamStutter(layoutForPerformance[xDisplay].paramKind, layoutForPerformance[xDisplay].paramID)) {
+					FXPress[xDisplay].padPressHeld = true;
+					FXPress[xDisplay].timeLastPadPress = AudioEngine::audioSampleTimer;
+
+					backupXMLDefaultFXPress[xDisplay].padPressHeld = FXPress[xDisplay].padPressHeld;
+					backupXMLDefaultFXPress[xDisplay].timeLastPadPress = FXPress[xDisplay].timeLastPadPress;
+				}
+			}
+		}
+		if (FXPress[xDisplay].padPressHeld) {
```

Comment:
This can break if the XML tags are out of order which is probably not what we want -- instead, we should automatically set `padPressHeld` if any of the relevant tags appear.

Reply:
> This can break if the XML tags are out of order which is probably not what we want -- instead, we should automatically set `padPressHeld` if any of the relevant tags appear.

Not sure I understand. The relevant tags can be there but the status can be set to "Off" - in this case we would not want to set it to held.

If you're referring to line 1976, if (FXPress[xDisplay].padPressHeld), padPressHeld is initialized to false, so this condition would only ever be true if the <status> tag is found. 

is your suggestion simply to just remove the if condition and move the code within that if to below line 1972?

Reply:
My suggestion would be to remove the `if` condition and unconditionally parse these tags. If that would cause some problems, once the full fxpress is parsed (i.e. after the loop) we can apply some fixup logic (e.g. resetting these back to their defaults if `!fxPress[xDisplay].padPressHeld`

Reply:
Ah I understand what you mean - you meant that potentially <row> and <resetValue> wouldn't be parsed. Ok I know what to do, thanks!

Reply:
@bobtwinkles just made the changes, please let me know if I understood correctly! Thanks :)